src/jit/morph.cpp

```diff
@@ -11416,6 +11416,20 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                 }
             }
 
+            // If gtOp1 is a GT_FIELD, we need to pass down the mac if
+            // its parent is GT_ADDR, since the address of the field
+            // is part of an ongoing address computation. Otherwise
+            // op1 represents the value of the field and so any address
+            // calculations it does are in a new context.
+            if ((op1->gtOper == GT_FIELD) && (tree->gtOper != GT_ADDR))
```

Comment:
I'll probably leave it as is because it is kind of subtle (maybe this is what Brian was alluding to).

We can change this check as you suggest, but if we also do likewise for the similar `op2` check, which seems like an entirely plausible thing to do, we could miss the need to null check an (ADD, FIELD, FIELD) case.

Probably more straightforward to account for the impact of `op1` on any ongoing address computation when we're looking at `op1` (and similarly for `op2`) rather than the swapped checking we do here.

Also not clear to me why other binary operators aren't considered here.